### PR TITLE
Filters provided by filter providers are made to never be cached

### LINKED ISSUE #4504: Possible double-execution of filter providers
We're seeing a filter provider called twice per-request in an orchard scenario

Stacks:

### PR DESCRIPTION
**Pull request by kichalla**
[Fixes #4504] Possible double-execution of filter providers

@rynowak @sebastienros

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 1]
```diff
@@ -85,7 +91,7 @@ public Entry GetCacheEntry(ControllerContext controllerContext)
             }
             else
             {
-                entry = new Entry(null, items, executor);
+                entry = new Entry(null, staticallyDefinedFilterItems, executor);
             }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: This logic, along with `allFiltersCached` needs to go away. We just want to store the items. This is a bit of a step back in perf, but we need to call the providers every request.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 2]
```diff
@@ -56,22 +56,28 @@ public Entry GetCacheEntry(ControllerContext controllerContext)
                 return entry;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: Which method is called by the invoker? For instance calling `GetFilters` can still double-execute the providers. I think we've got more cleanup to do here.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterItemOrderComparer.cs [hunk 3]
```diff
@@ -1,34 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using System;
-using System.Collections.Generic;
-using Microsoft.AspNetCore.Mvc.Filters;
-
-namespace Microsoft.AspNetCore.Mvc.Internal
-{
-    public class FilterItemOrderComparer : IComparer<FilterItem>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterItemOrderComparer.cs]: Was this unused?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: `GetFilters` and `GetControllerActionMethodExecutor` should we deleted. We do not want to look up the cache entry more than once per request.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs [hunk 4]
```diff
@@ -186,8 +186,80 @@ public void GetControllerActionMethodExecutor_CachesExecutor()
             Assert.Same(executor1, executor2);
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void GetFilters_FiltersFromFilterProviders_AreNeverCached(bool reusable)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs]: `GetFilters_FiltersAddedByAFilterProvider_AreNeverCached`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs [hunk 5]
```diff
@@ -186,8 +186,80 @@ public void GetControllerActionMethodExecutor_CachesExecutor()
             Assert.Same(executor1, executor2);
         }
 
+        [Theory]
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs]: I'd also like to see a test that verifies how many times a filter provider is called.

**Comment by rynowak** [comment]
:watch:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs (hunk 5 above)
Comment (kichalla) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs]: This test covers the scenario. The filter provider adds a new filter item always to the result. If there were duplicate items the count of item would be incorrect.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterItemOrderComparer.cs (hunk 3 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterItemOrderComparer.cs]: yes right

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: > but we need to call the providers every request.

The providers are getting called for every request via the `GetFiltersFromEntry` method.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterActionInvoker.cs#L126-L127

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerCacheTest.cs]: Right, but this is a unit test for a method that isn't called by product code :trollface:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: Which means that on the first request they are called twice.

**Comment by kichalla** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 6]
```diff
@@ -45,96 +45,59 @@ private InnerCache CurrentCache
...
-                entry = new Entry(filters, null, executor);
-            }
-            else
-            {
-                entry = new Entry(null, items, executor);
-            }
-
-            cache.Entries.TryAdd(actionDescriptor, entry);
-            return entry;
+            return new EntryInfo(allFilters, cacheEntry);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: This is looking good!

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 7]
```diff
@@ -186,25 +147,38 @@ public InnerCache(int version)
...
                 ActionMethodExecutor = executor;
             }
-            public IFilterMetadata[] Filters { get; }
 
             public List<FilterItem> FilterItems { get; }
 
             public ObjectMethodExecutor ActionMethodExecutor { get; }
         }
+
+        public struct EntryInfo
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: Let's call this something else ... `ControllerActionInvokerState`?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 8]
```diff
@@ -186,25 +147,38 @@ public InnerCache(int version)
...
+                IFilterMetadata[] allFilters,
+                Entry cacheEntry)
+            {
+                AllFilters = allFilters;
+                CacheEntry = cacheEntry;
+            }
+
+            public IFilterMetadata[] AllFilters { get; }
+
+            public Entry CacheEntry { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: We don't need the cache entry here, just the objectmethodexecutor

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 9]
```diff
@@ -186,25 +147,38 @@ public InnerCache(int version)
...
+        {
+            public EntryInfo(
+                IFilterMetadata[] allFilters,
+                Entry cacheEntry)
+            {
+                AllFilters = allFilters;
+                CacheEntry = cacheEntry;
+            }
+
+            public IFilterMetadata[] AllFilters { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: Just call this `Filters`

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs [hunk 10]
```diff
@@ -186,25 +147,38 @@ public InnerCache(int version)
...
             }
 
-            public ConcurrentDictionary<ActionDescriptor, Entry> Entries { get; } = 
+            public ConcurrentDictionary<ActionDescriptor, Entry> Entries { get; } =
                 new ConcurrentDictionary<ActionDescriptor, Entry>();
 
             public int Version { get; }
         }
 
         public struct Entry
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: This doesn't need to be public anymore

**Comment by rynowak** [comment]
:watch: for minor feedback, looking good overall

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs (hunk 7 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs]: Sure

**Comment by kichalla** [comment]
:up: :date:

**Comment by rynowak** [comment]
:shipit: - do you know if this is approved for RC2?

**Comment by kichalla** [comment]
@rynowak actually i haven't checked. will do now.

**Comment by kichalla** [comment]
a259638d79ed5c02b8a2805234c93dd83d739cf5

**Pull request closed without merging**